lib/prelude.dx

```diff
@@ -1516,7 +1520,83 @@ def tail (xs:n=>a) (start:Int) : List a =
   numElts = size n - start
   toList $ slice xs start (Fin numElts)
 
---TODO: move this to be with other random number functions
+
+'## Pseudorandom number generator utilities
+Dex does not use a stateful random number generator.
+Rather it uses what is known as a split-able random number generator, which is based on a hash function.
+Dex's PRNG system is modelled directly after [JAX's](https://github.com/google/jax/blob/master/design_notes/prng.md), which is based on a well established but shockingly underused idea from the functional programming community: the splittable PRNG. It's a good idea for many reasons, but it's especially helpful in a parallel setting. If you want to read more, [Splittable pseudorandom number generators using cryptographic hashing](http://publications.lib.chalmers.se/records/fulltext/183348/local_183348.pdf) describes the splitting model itself and [D.E. Shaw Research's counter-based PRNG](http://www.thesalmons.org/john/random123/papers/random123sc11.pdf) proposes the particular hash function we use.
+
+'### Key functions
+
+-- TODO: newtype
+Key = Word64
+
+def threeFry2x32 (k:Word64) (count:Word64) : Word64 =
+  -- Based on jax's threefry_2x32 by Matt Johnson and Peter Hawkins
+  rotations1 = [13, 15, 26, 6]
+  rotations2 = [17, 29, 16, 24]
+
+  k0 = lowWord k
+  k1 = highWord k
+  -- TODO: add a fromHex
+  k2 = k0 .^. k1 .^. (IToW32 466688986) -- 0x1BD11BDA
+
+  x = lowWord count
+  y = highWord count
+  x = x + k0
+  y = y + k1
+
+  rotations = [rotations1, rotations2]
+  ks = [k1, k2, k0]
+  (x, y) = yieldState (x, y) \ref. for i:(Fin 5).
+    (x, y) = yieldState (get ref) \ref. for j.
```

Comment:
Is there any good reason to use nested state here? You should be able to just put `for j` here and use the `ref` from the outside. Or am I missing something?

Reply:
Nope, my bad! I removed the nested state.